### ISSUE TITLE
Hash with symbol keys using shorthand notation will be interpreted as NamedTuple

### ISSUE THREAD

**Issue by noraj** [crystal-lang/crystal#13268]
## Discussion

I would like to discuss an edge case around the usage of `Hash` vs `NamedTuple`.

About NamedTuple:

> A named tuple is a fixed-size, immutable, stack-allocated mapping of a fixed set of keys to values.
>
> You can think of a [NamedTuple](https://crystal-lang.org/api/1.7.3/NamedTuple.html) as an immutable [Hash](https://crystal-lang.org/api/1.7.3/Hash.html) whose keys (which are of type [Symbol](https://crystal-lang.org/api/1.7.3/Symbol.html)), and the types for each key, are known at compile time.

The issue is that initializing a **hash** with symbol keys and constant values it is then instantiated as a NamedTuple instead of a Hash. Eg.

```crystal
just = {
  type: 4,
  laddr: 0,
  raddr: 0,
  state: 0
}
```

So when latter I want to update a value of the Hash with `just[:type] = a_method(x)` and error is thrown because NamedTuples are immutable.

```
Error: undefined method '[]=' for NamedTuple(type: Int32, laddr: Int32, raddr: Int32, state: Int32)
```

This comes from the confusion and the collapsing syntax between the two classes, it's not possible to determine if an object is a Hash or NamedTuple when the Hash keys are symbols and values are statically initialized.

I think NamedTuple should have been using a different syntax from the start to avoid this kind of conflicts. But now to avoid a breaking change an option could be provided to explicitly tell the Hash is a Hash and not a NamedTuple.

As a bad workaround one could initialize the Hash with `Hash.new` and then allocated some values. A cleaner workaround is to do `a = { key: value }.to_h` to ensure if it's detected as a NamedTuple it's then converted as a Hash. Another workaround a bit more annoying workaround is to use the `:a => b` syntax instead of `a: b`.

Here are test cases, however, I have no idea in mind for unpredictable type at compile type to test if the `a: b` syntax will lead to Hash or NamedTuple in such a case.

```crystal
require "json"

a = { "test" => 5 }
puts typeof(a) # Hash because key is a string
b = { :test => 5 }
puts typeof(b) # Hash because it uses the :a => b syntax instead of a: b
c = { test: JSON.parse(File.read("/tmp/test.json"))["a"] }
puts typeof(c) # NamedTuple because dispite the type being impredicatble at compile time it cheats with a `JSON::Any` type.
d = { test: 5 }
puts typeof(d) # NamedTuple because the type is known at compile time
```

In the hypothesis that the classical symbol notation `:a => b` would be strictly reserved for `Hash` and the modern symbol notation `a: b` would be strictly reserved for NamedTuple and no plans are considered to change any actual behavior, then I suggest updating the following pages of the documentation to explicitly state it.

- https://crystal-lang.org/reference/1.7/crystal_for_rubyists/index.html#types -> add a `Hash` section to tell that `a: b` will be seen as NamedTuple (Symbol shorthand notation for Hashes not supported in Crystal vs Ruby) + offer solutions
- https://crystal-lang.org/reference/1.7/syntax_and_semantics/literals/hash.html -> same warning
- https://crystal-lang.org/reference/1.7/syntax_and_semantics/literals/named_tuple.html -> same warning

Tell me if the assumption is correct so I can update the doc.

**Comment by devnote-dev** [comment]
> The issue is that initializing a hash with symbol keys and constant values it is then instantiated as a NamedTuple instead of a Hash.
> [...]

I tried the following code, it works as intended:

```crystal
thing = {
  :foo => "foo",
  :bar => "bar"
}
puts typeof(thing) # => Hash(Symbol, String)

thing[:baz] = "baz"
puts thing # => {:foo => "foo", :bar => "bar", :baz => "baz"}
```

Also note that the syntax in your example is named tuple syntax, not hash syntax which uses `=>`.

> Here are test cases, however, I have no idea in mind for unpredictable type at compile type to test if the a: b syntax will lead to Hash or NamedTuple in such a case.

Example A and B are hashes because you're using `=>`, they're types are still known at compile time but it is not as strict as a named tuple, which C and D are because of `:`.

Having warnings in the documentation would be useful though, it's not exactly obvious that hashes use a specific syntax which can be confusing if you're coming from Ruby (or JavaScript, Python, etc).

**Comment by Blacksmoke16** [comment]
Yes, as @devnote-dev pointed out, there is no "shorthand notation". Using a `:` denotes a named tuple while `=>` denotes a hash. This is somewhat pointed out in the hash docs:

https://crystal-lang.org/reference/1.7/syntax_and_semantics/literals/hash.html

> Hashes are typically created with a hash literal denoted by curly braces ({ }) enclosing a list of pairs using => as delimiter between key and value and separated by commas ,.

NamedTuple literal docs don't directly point out that you use `:`, but its part of the example.

I'm hesitant to think we need to document these "warnings" explicitly in the literal docs for each type. Having something more descriptive for named tuples like we have for hashes would probably be sufficient. Adding an explicit section in the `Crystal for Rubyists` would be a better alternative as this page fits this topic more.

Going to transfer this issue to the `crystal-book` repo.
